examples/06_deduplication.py

```diff
@@ -0,0 +1,187 @@
+"""
+Deduplicating misspelled categories with deduplicate
+====================================================
+
+A common step in data analyses is grouping or analyzing data conditional on a
+categorical variable. In real world datasets, there often will be slight
+misspellings in the category names: This happens when, for example, data input
+*should* use a drop down menu, but users are forced to input the category name
+by hand. Misspellings happen and analyzing the resulting data using a simple
+`GROUP BY` is not possible anymore.
+
+This problem is however the perfect use case of *unsupervised learning*, a
+category of various statical methods that find structure in data without
+providing explicit labels/categories of the data a-priori. Specifically
+clustering of the distance between strings can be used to find clusters
+of strings that are similar to each other (e.g. differ only by a misspelling)
+and hence gives us an easy tool to flag potentially misspelled category names
+in an unsupervised manner.
+"""
+
+###############################################################################
+# An example
+# ----------
+#
+# Imagine the following example:
+# As a data scientist, our job is to analyze the data from a hospital ward.
+# We notice that most of the cases involve the prescription of one of three different medications:
+#  "Contrivan", "Genericon", or "Zipholan".
+# However, data entry is manual and - either because the prescribing doctor's handwriting
+# was hard to decipher, or due to mistakes during data input - there are multiple
+# spelling mistakes for these three medications.
+#
+# Let's generate some example data that demonstrate this.
+
+
+import numpy as np
+import pandas as pd
+import matplotlib.pyplot as plt
+
+
+def generate_example_data(examples, entries_per_example, prob_mistake_per_letter):
+    """Helper function to generate data consisting of multiple entries per example.
+    Characters are misspelled with probability `prob_mistake_per_letter`"""
+    import string
+
+    data = []
+    for example, n_ex in zip(examples, entries_per_example):
+        len_ex = len(example)
+        # generate a 2D array of chars of size (n_ex, len_ex)
+        str_as_list = np.array([list(example)] * n_ex)
+        # randomly choose which characters are misspelled
+        idxes = np.where(
+            np.random.random(len(example[0]) * n_ex) < prob_mistake_per_letter
+        )[0]
+        # and randomly pick with which character to replace
+        replacements = [
+            string.ascii_lowercase[i]
+            for i in np.random.choice(np.arange(26), len(idxes)).astype(int)
+        ]
+        # introduce spelling mistakes at right examples and right char locations per example
+        str_as_list[idxes // len_ex, idxes % len_ex] = replacements
+        # go back to 1d array of strings
+        data.append(np.ascontiguousarray(str_as_list).view(f"U{len_ex}").ravel())
+    return np.concatenate(data)
```

Comment:
Best is to move this function into [datasets._fetching](https://github.com/dirty-cat/dirty_cat/blob/main/dirty_cat/datasets/_fetching.py) and rename it to something like `make_deduplication` so that we use it whenever we want to generate an example for this.
And to have a one-line import rather than a function makes the important features of this example much more visible.

Note: I see you used the same function in the [tests](https://github.com/dirty-cat/dirty_cat/blob/09a88c4053a031965ec8f08c6331491cceec5a51/dirty_cat/tests/test_deduplicate.py#L18), so maybe import it there as well